bazel/pgv_proto_library.bzl

```diff
@@ -10,12 +10,78 @@ def pgv_go_proto_library(name, srcs = None, deps = [], **kwargs):
                      validate = 1,
                      **kwargs)
 
-def pgv_cc_proto_library(name, srcs = None, deps = [], **kwargs):
-    cc_proto_library(name,
-                     srcs,
-                     protoc = "@com_google_protobuf//:protoc",
-                     default_runtime = "@com_google_protobuf//:protobuf",
-                     deps = ["//validate:validate_cc"] + deps,
-                     visibility = ["//visibility:public"],
-                     validate = 1,
-                     **kwargs)
+def _CcValidateHdrs(srcs):
+  ret = [s[:-len(".proto")] + ".pb.validate.h" for s in srcs]
+  return ret
+
+def pgv_cc_proto_library(
+        name,
+        srcs=[],
+        deps=[],
+        cc_libs=[],
+        include=None,
+        protoc="@com_google_protobuf//:protoc",
+        protoc_gen_validate = "//:protoc-gen-validate",
+        internal_bootstrap_hack=False,
+        use_grpc_plugin=False,
+        default_runtime="@com_google_protobuf//:protobuf",
+        **kargs):
+  """Bazel rule to create a C++ protobuf validation library from proto source files
+
+  Args:
+    name: the name of the pgv_cc_proto_library.
+    srcs: the .proto files of the pgv_cc_proto_library.
+    deps: a list of dependency labels; must be cc_proto_library.
+    include: a string indicating the include path of the .proto files.
+    protoc: the label of the protocol compiler to generate the sources.
+    protoc_gen_validate: override the default version of protoc_gen_validate.
+                   Most users won't need this.
+    default_runtime: the implicitly default runtime which will be depended on by
+        the generated cc_library target.
+    **kargs: other keyword arguments that are passed to cc_library.
+
+  """
+
+  # Generate the C++ protos
+  cc_proto_library(
```

Comment:
This is a nice improvement. Do you know if there a way to do this with `native.cc_proto_library`, `native.proto_library` and Bazel toolchains? That's where we want to go long term (but this might be a good stepping stone).

Reply:
You can define toolchains, but apparently the functionality to invoke them is [not exposed yet](https://groups.google.com/d/msg/bazel-discuss/xoy98a4yhAA/28xEMZQ-CQAJ).

Reply:
It's worth noting that the `native.cc_proto_library` is not a drop-in replacement for the current `cc_proto_library` since it expects a `native.proto_library` target as a dependency instead of a list of proto source files. Not a huge issue, but it'll break some things when we do switch.

Reply:
I think it's only not exposed to Skylark rules, but it does apply to the Bazel native `{cc,java}_proto_library`. I agree that it's not a drop-in replacement, it's actually a bit better than the way the Protobuf `cc_proto_library` operates, due to its separation of concerns with `proto_library`. We can merge this as it passes tests and is strictly better, I can take a look later this week at whether we can make toolchains work.